File: PythonInterviewQuestions/spiralList.py
def displayList(a):
    for i in range(len(a)):
        for j in range(len(a[i])):
            print(a[i][j], end=" ")
        print()


def generateList(n, length):
    L = []
    for i in range(1, n+1, length):
        A = []
        for j in range(5):
            A.append(i+j)
        L.append(A)

    return L


def getSpiralList(A):
    sr, er = 0, len(A)
    sc, ec = 0, len(A[0])
    L = []

    while sc < ec:
        for i in range(sc, ec):
            L.append(A[sr][i])
        sr += 1
        if sr == er:
            break

        for i in range(sr, er):
            L.append(A[i][ec-1])
        ec -= 1
        if sc == ec:
            break

        for i in range(ec-1, sc-1, -1):
            L.append(A[er-1][i])
        er -= 1
        if sr == er:
            break

        for i in range(er-1, sr-1, -1):
            L.append(A[i][sc])
        sc += 1

    return L


def main():
    L = generateList(25, 5)
    displayList(L)
    print(getSpiralList(L))


if __name__ == '__main__':
    main()

File: PerlInterviewQuestions/spiralList.pl
use strict;
use warnings;
use Data::Dumper;

sub displayList {
    my $L = shift();

    for (my $i = 0; $i < scalar(@$L); $i++) {
        for (my $j = 0; $j < scalar(@{$$L[$i]}); $j++) {
            print $$L[$i][$j] . " ";
        }
        print "\n";
    }
}

sub generateList {
    my $n = shift;
    my $length = shift;
    my @L = ();
    for (my $i = 1; $i <= $n; $i += $length) {
        my @A = ();
        for (my $j = $i; $j < $i + $length; $j++) {
            push(@A, $j);
        }
        push(@L, \@A);
    }
    return @L;
}

sub getSpiralList {
    my $L = shift();
    my @S = ();
    my $sr = 0;
    my $sc = 0;
    my $er = scalar @$L;
    my $ec = scalar @{$$L[0]};

    while ($sc < $ec) {
        for (my $i = $sc; $i < $ec; $i++) {
            push(@S, $$L[$sr][$i]);
        }
        $sr++;
        last if ($sr == $er);

        for (my $i = $sr; $i < $er; $i++) {
            push(@S, $$L[$i][$ec - 1]);
        }
        $ec--;
        last if ($sc == $ec);

        for (my $i = $ec - 1; $i > $sc - 1; $i--) {
            push(@S, $$L[$er - 1][$i]);
        }
        $er--;
        last if ($sr == $er);

        for (my $i = $er - 1; $i > $sr - 1; $i--) {
            push(@S, $$L[$i][$sc]);
        }
        $sc++;
    }

    return(@S);
}

my @L = generateList(20, 5);
$Data::Dumper::Indent = 0;
print Dumper \@L;
print "\n";
displayList(\@L);
print "\n";
my @spiralList = getSpiralList(\@L);
print Dumper \@spiralList;

